cloudstack_plugin/volume.py

```diff
@@ -0,0 +1,129 @@
+########
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# * See the License for the specific language governing permissions and
+# * limitations under the License.
+
+import copy
+from cloudify import ctx
+from cloudify.decorators import operation
+from cloudify.exceptions import NonRecoverableError
+
+from cloudstack_plugin.cloudstack_common import (
+    get_cloud_driver,
+    get_resource_id,
+    CLOUDSTACK_ID_PROPERTY,
+    CLOUDSTACK_NAME_PROPERTY,
+    CLOUDSTACK_TYPE_PROPERTY,
+    COMMON_RUNTIME_PROPERTIES_KEYS,
+    USE_EXTERNAL_RESOURCE_PROPERTY,
+)
+
+VOLUME_CLOUDSTACK_TYPE = 'volume'
+
+RUNTIME_PROPERTIES_KEYS = COMMON_RUNTIME_PROPERTIES_KEYS
+
+
+@operation
+def create(**kwargs):
+    """ Create a volume
+    """
+
+    cloud_driver = get_cloud_driver(ctx)
+    volume = {}
+
+    if ctx.node.properties[USE_EXTERNAL_RESOURCE_PROPERTY] is False:
+
+        ctx.logger.debug('reading volume attributes.')
+        volume.update(copy.deepcopy(ctx.node.properties['volume']))
+
+        if 'name' in volume:
+            volume_name = volume['name']
+        else:
+            raise NonRecoverableError("To create a volume, the name of the "
+                                      "volume is needed")
+
+        if 'size' in volume:
+            volume_size = volume['size']
+        else:
+            raise NonRecoverableError("To create a volume, the size of the "
+                                      "volume is needed")
+
+        volume = cloud_driver.create_volume(name=volume_name,
+                                            size=volume_size)
+
+        if volume_exists(cloud_driver, volume.id):
+
+            ctx.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY] = volume.id
+            ctx.instance.runtime_properties[CLOUDSTACK_TYPE_PROPERTY] = \
+                VOLUME_CLOUDSTACK_TYPE
+
+        else:
+            raise NonRecoverableError("Volume not created")
+
+    elif ctx.node.properties[USE_EXTERNAL_RESOURCE_PROPERTY] is True:
+
+        if ctx.node.properties['resource_id']:
+            resource_id = ctx.node.properties['resource_id']
+
+            volume = get_volume_by_id(cloud_driver, resource_id)
```

Comment:
if a user specified a faulty resource_id this code would not detect it. get_volume_by_id should either raise an exception if the volume could not be found, or this code should check for None.